Experiment Contextual onboarding

alessandroboron opened the pull request:
Task/Issue URL: https://app.asana.com/0/72649045549333/1207585462837397/f

**Description**:

Add an experiment to show an in-context onboarding to iOS.

**NOTE**
This PR does not yet contain the localisation of the strings and new / updated pixels. 
The plan is to have two small PRs for those tasks that can be merged into this one.

**NOTE 2**
Everything has been reviewed incrementally. Although I added some commits in the last day to fix some issues. I’m referring to commits added after [Contextual Onboarding Ship] Review(https://github.com/duckduckgo/iOS/pull/3160/commits/9d492fe1754414c86adb9dd75deb591a6efc7408)

**Steps to test this PR**:
I’ve added [Testing Scenarios](https://app.asana.com/0/1206329551987282/1207912446107172/f) for the new flow. 
To test the scenarios you can add the hardcoded value  `return VariantIOS(name: "mb", weight: 1, isIncluded: VariantIOS.When.always, features: [.newOnboardingIntro])` in `DefaultVariantManager` at line 144.

If you have any doubts please ping me. Additionally I leave here these resources: 
[Trigger & Dismissal logic](https://app.asana.com/0/72649045549333/1207816658301674/f)
[Android User Journey Map](https://app.asana.com/app/asana/-/get_asset?asset_id=1207856243262446) 

**Definition of Done (Internal Only)**:

* [x] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 15
* [ ] iOS 16
* [ ] iOS 17

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

SabrinaTardio committed:
implement contextual dialogs (#3039)

Task/Issue URL: https://app.asana.com/0/1204186595873227/1207736729379245/f

Description: Implements the Dialog we will use in the contextual onboarding

SabrinaTardio committed:
add sites providers (#3055)

Task/Issue URL: https://app.asana.com/0/1204186595873227/1207569384066479/f

Description: Implements the providers for search and site suggestions showed in the contextual onboarding according to https://app.asana.com/0/1202552961248957/1207407312592957/f

SabrinaTardio committed:
new tab onboarding (#3080)

Task/Issue URL: https://app.asana.com/0/1204186595873227/1207569384066444/f

Description: Implements new tab page messages for the new onboarding. The new or old messages will shown depending on the variant manager

SabrinaTardio committed:
fix nt search details (#3090)

Task/Issue URL: https://app.asana.com/0/0/1207812824983895/f

Description: Avoids showing https in the search query and avoids “” in the search query

alessandroboron committed:
Onboarding Contextual Tutorial Rendering (#3063)

alessandroboron committed:
Contextual Onboarding Linear (#3089)

Co-authored-by: Sabrina Tardio <44158575+SabrinaTardio@users.noreply.github.com>

alessandroboron committed:
Contextual Onboarding - Show final dialog (#3108)

Task/Issue URL: https://app.asana.com/0/1206329551987282/1207701571713573/f

Description:

Show in context final dialog
Set flag when final dialog is shown by NTP and use the flag to decide if we should show the final dialog in NTP.
Ensure Fire Dialog is not show multiple times.
Make Tracker dialogs have more priority than final dialog

SabrinaTardio committed:
Contextual Onboarding dialog improvements (#3105)

Co-authored-by: Alessandro Boron <aboron@duckduckgo.com>

alessandroboron committed:
Fix tests due to new addition to main

alessandroboron committed:
Update Sites suggestion (#3131)

Task/Issue URL: https://app.asana.com/0/1206329551987282/1207874379629922/f

Description:

Updates sites suggestion according to ticket.

alessandroboron committed:
Update German Surprise me to duden.de

alessandroboron committed:
Contextual Onboarding - Privacy Icon animation (#3130)

SabrinaTardio committed:
Alessandro/refresh dax dialog and ntp logo (#3129)

Co-authored-by: Alessandro Boron <aboron@duckduckgo.com>

SabrinaTardio committed:
Contextual Onboarding - Ensure hosted SwiftUI view updates container size for for iOS < 16 (#3137)

Co-authored-by: Alessandro Boron <aboron@duckduckgo.com>

alessandroboron committed:
Fix unit test compilation due to conflicts

alessandroboron committed:
Contextual Onboarding Ship Review (#3149)

Co-authored-by: Sabrina Tardio <44158575+SabrinaTardio@users.noreply.github.com>

alessandroboron committed:
Mitigate fire button pulse animation appearing when it should not for new onboarding

alessandroboron committed:
Do not show Fire button education message when tapping FireButton on iPad

alessandroboron committed:
Fix an issue where search suggestions would show if the user skipped them and seen the fire dialog.

alessandroboron committed:
Add new fire dialog setting to avoid old code to change it

alessandroboron committed:
Fix unit tests compilation due to rebase

alessandroboron committed:
Fix Dax Dialog not reloading properly when going back to the view after animating title changes

alessandroboron committed:
Implement custom ListButtonStyle and ensure list buttons fits two lines

alessandroboron committed:
Fix animation starting when the view is not on screen yet.

alessandroboron committed:
Make Fire button pulse have priority on privacy button pulse

alessandroboron committed:
Revert isEnabled changes

alessandroboron committed:
Update Pixel to avoid missing the pixel name. Work on Pixel will be done in another PR

alessandroboron committed:
Fix Swiftlint Issue

brindy reviewed (dismissed):
Looking good, only issue I've been able to find is that it's possible to trigger the old dax dialog for adding favorites via the widget.  Though it's not clear what we should do in this situation.  We can't really change the widget (because they are not variant aware) so maybe we just don't show old dax in this case.

![dax-add-favorites-flow](https://github.com/user-attachments/assets/e6daa612-efd9-4bb4-adc0-21a5954633b9)

![favorites-widget](https://github.com/user-attachments/assets/f67c102b-c1c0-43a6-828b-357ee03e59b6)

alessandroboron committed:
Contextual Onboarding - Disable Add Favorites flow for experiment group (#3167)

Task/Issue URL: https://app.asana.com/0/1206329551987282/1207937793719014/f

**Description**:
1. Disable favorites flow for experiment cohort variant/
2. Fix Dax Logo overlap with favorites when the users finish the contextual onboarding.
3. Ignore deep links into the App via a widget if the user is doing Onboarding intro.

alessandroboron committed:
Contextual Onboarding Pixels (#3177)

Task/Issue URL: https://app.asana.com/0/1206329551987282/1207569383033366/f
**Description**:
Adds Pixels for Contextual Onboarding.

alessandroboron committed:
Contextual Onboarding - Translations (#3180)

Task/Issue URL: https://app.asana.com/0/414235014887631/1207556941490012/f

**Description**:
Add translations for contextual onboarding.

alessandroboron committed:
Fix issue where showing a system permission would trigger fire pulse when privacy pulse was on

samsymons reviewed (approved):
LGTM, thanks for addressing the issue with didBecomeActive that we discussed.

There's only one more issue like that that I can see, which you can reproduce by entering the Fire button onboarding state with the highlight ring animating, then background/foreground the app - when you return, the highlight ring is gone.

Other than that, I've tested on different devices, device orientations etc., all seemed good from what I can tell.
